rfc/src/rfcs/0003-resource-limits-timeouts.md

```diff
@@ -0,0 +1,224 @@
+* "Typical" machines for quite awhile were 4 core 8 GB memory. Lately there is more variety but 8 core 16 GB seems common, which is also a 1:2.
+* In the Kani test suite, most `cbmc` processes used about 1 GB or less. A small number hit 3-4 GB.
+* CBMC often keeps "virtual memory" and "resident set size" in pretty close alignment.
+
+The primary goal of a memory limit is making a parallel harness runner reliable.
+Enforcing the memory limits means we can plan on parallel processes not exceeding available machine memory.
+
+The advantages of a lower default memory limit are:
+
+1. We could still achieve a high level of parallelism even in the presence of higher-limit harnesses.
+(Workaround: Explicitly setting a lower limit than default would allow this too!)
+
+The advantages of a higher default memory limit are:
+
+1. Fewer harnesses will need an annotation, allowing more users to be unbothered by the limit.
+
+Given these trade-offs, and the data on common CI machine sizes, I think a default memory limit of **2 GB** seems best.
+This should be the largest value that would still allow maximum parallelism on common machine types.
+We may want to collect data on real-world impact to customer harnesses.
+
+### Question 9: What choice should we make for the default timeout?
+
+Here the data we would like to examine is less about machines and more about common proofs.
+
+The advantages of a higher default timeout are:
+
+1. Fewer harnesses would be impacted by needing an explicit limit.
+
+The advantages of a lower default timeout are:
+
+1. Customers experiencing non-termination will get faster feedback from our tools.
+2. Solving "plans" will be able to more accurately forecast how long each harness will take to solve.
+3. Parallel runners can prioritize long-running harness, to ensure high CPU utilizaiton throughout verification.
+
+We may need more data to make a final decision here, but I'd suggest we bias towards small: **15 seconds** default timeout.
+
+### Question 10: Do we need more than per-harness timeouts?
+
+No.
+
+We could decide to add a over-all timeout (encompassing build and link times), but I argue this is not helpful.
+The timeout we really care about (where nontermination exists and affects users) is for running `cbmc` on harnesses.
+
+## Roll-out plan for this breaking change
+
+Application of default limits will be a breaking change.
+I propose the following approach:
+
+1. Introduce limits and `--ignore-limits` but make this option do nothing at first: it is always applied.
+This replicates the current behavior with respect to enforcing limits, but will cause warnings to be issued with suggested limit settings.
+2. Keep this as the default for 2 kani releases (4 weeks).
+Perhaps longer if issues arise.
+3. Flip the default to enforcing.
+This will finally cause the breaking change, but will have had some warning, and we retain an option to revert to the old behavior.
+
+## Open questions
+
+TBD
+
+## Future possibilities
```

Comment:
A few ideas that came to mind as future possibilities:

- Could we add a way to set a machine-independent resource limit? [This thread](https://github.com/Z3Prover/z3/issues/56) has a pretty good discussion of why this can be useful for reproducible verification results. Machine-independent resource limits are part of the SMT-LIB standard and [have been proposed for CBMC](https://github.com/diffblue/cbmc/issues/4452). This would help ensure that harnesses don't time out in CI when they work fine locally.
- Could Kani provide some sort of functionality where it runs a harness (perhaps a couple of trials), and then adds resource bound annotations to the harness based off of the actual amount of resources it takes to complete? This could help catch regressions in prover performance. (Perhaps it'd be more useful for Kani internally than for clients.)

Reply:
GREAT ideas. This also brings to mind the mechanism I proposed to implement the limits currently is pretty primitive and will need adjusting if we start fully supporting external solvers (vs just minisat inside the cbmc process).